### ISSUE TITLE
Support CIDR function or allow custom functions

### ISSUE THREAD

**Issue by sporkmonger** [dgraph-io/dgraph#1556]
I'm working on a security related project and I'm interested in being able to query IP addresses by CIDR range. The CIDR ranges themselves may also be attributes. My first thought was to create a custom `cidr` query function, but it appears that there's no capability to add custom functions. I don't know how useful a `cidr` query function would be to anyone else, but if others thought it would be useful, I'd be happy to implement a patch. If not, I'd be interested in seeing a mechanism for implementing custom functions so that I could have an efficient mechanism for querying IPs that wouldn't require running a fork.

**Comment by manishrjain** [comment]
Would be good to provide an interface for our tokenizers, which can be coded up to support custom indexers.

**Comment by peterstace** [comment]
I'm doing some experimentation with https://golang.org/pkg/plugin/ to integrate custom tokenizers into Dgraph. I've got a basic prototype working.

Users who want custom tokenizers will have to write the tokenization code, compile it into an `.so` file, then pass it as an arg to Dgraph when it starts up. Upside: 1) users don't have to run forks. 2) Users don't have to compile dgraph, they can just use the stock binaries. Downside: Only works on Linux (for now -- support for the `plugin` package may be added to other platforms in the future).

**Comment by peterstace** [comment]
How this works with CIDR example:

User writes some custom functionality in a go file. It has to be a `main` package, detail of how this works here: https://golang.org/pkg/plugin/

The symbols that Dgraph hooks onto are `Name` and `Tokens`. The types must be `string` and `func(string) ([]string, error)`.

```
package main                                                                                                                                  

import (
    "net"
)

var Name string = "cidr"

func Tokens(s string) ([]string, error) {
    _, network, err := net.ParseCIDR(s)
    if err != nil {
        return nil, err 
    }   
    return []string{network.String()}, nil 
}

```

The user can compile with `go build -buildmode=plugin main.go`.

Then, the user runs Dgraph passing in `--custom_tokenizer=main.so`. These can be comma separated, allowing multiple plugins to run at the same time.

When creating the schema, we now have access to the `cidr` tokenizer.

```
mutation{
  schema{
    ip: string @index(cidr) .
  }
  set {
    _:a <ip> "192.168.0.4/16" .
    _:b <ip> "192.168.0.5/16" .
    _:c <ip> "192.170.0.1/16" .
    _:d <ip> "192.170.0.22/16" .
    _:e <ip> "192.170.1.1/16" .
  }
}
```

Now, we can use `customallof` and `customanyof` to search/filter using the index. In this case, the tokenizer is a 1-1 mapping (CIDR form IP address to CIDR subnet), so it doesn't matter which one we use. I'll give another example later where `customallof` vs `customanyof` makes a difference.

```
{
  q(func: customanyof(ip, cidr, "192.168.0.0/16")) {
    ip
  }
}
```
```
{
  "data": {
    "q": [
      {
        "ip": "192.168.0.4/16"
      },
      {
        "ip": "192.168.0.5/16"
      }
    ]
  }
}
```
```
{
  q(func: customanyof(ip, cidr, "192.170.0.0/16")) {
    ip
  }
}
```
```
{
  "data": {
    "q": [
      {
        "ip": "192.170.1.1/16"
      },
      {
        "ip": "192.170.0.1/16"
      },
      {
        "ip": "192.170.0.22/16"
      }
    ]
  }
}
```

**Comment by peterstace** [comment]
@sporkmonger - I've finished getting a prototype for a plugin design working. Can you take a look and let me know if it will be suitable for your needs?

**Comment by sporkmonger** [comment]
Yes, sorry, had a busy few days on other stuff and haven't had a chance to
look yet.

**Comment by sporkmonger** [comment]
This looks like it's in the right ballpark. In practice, I'd have a dataset that's almost entirely, if not exclusively, /32s, but my queries would be /24s, /16s, etc.

**Comment by sporkmonger** [comment]
In terms of deployment, the Linux-only issue isn't a big problem provided that it's trivial to add custom plugins as a layer on top of the Docker image.

**Comment by sporkmonger** [comment]
FWIW, I think I'd prefer a query syntax that's more like `cidr(ip, "192.170.0.0/16")` since it's more approachable, but having looked at the parser, I get why that has issues.

**Comment by peterstace** [comment]
> In practice, I'd have a dataset that's almost entirely, if not exclusively, /32s, but my queries would be /24s, /16s, etc.

That's definitely possible with this plugin approach. You'd need a slightly more complex plugin than what I gave as an example though - it would need to produce multiple tokens per input, not just the one. E.g. Input is `"42.28.81.55/32"`, output would be `[]string{"42.28.81.0/24", "42.28.0.0/16", "42.0.0.0/8"}` (at whichever level of granularity you want the queries to work at).

> Linux-only issue isn't a big problem provided that it's trivial to add custom plugins as a layer on top of the Docker image.

That sounds like it should be straightforward. You'd just need to add the `.so` file to the docker image, and alter the command to invoke dgraph so that it knows where to read the file from. The only other caveat would be that the version of the go toolchain used to build the plugin has to be the same as the version used to build dgraph itself. There might be a way for dgraph to do that when it loads up the plugin.

> FWIW, I think I'd prefer a query syntax that's more like cidr(ip, "192.170.0.0/16") since it's more approachable, but having looked at the parser, I get why that has issues.

Yeah, the way I modified the query syntax was the most straightforward way I could find. Dynamically adding keywords in the parser depending on which plugins are loaded seemed like too complex of a solution.

With a syntax like `cidr(ip, "192.168.0.0/16")`, we'd also have to fit "any of" vs "all of" in there as well. The distinction doesn't really make sense for cidrs, but would for other types of tokenizers, e.g. a hypothetical tokenizer that creates 1 token per UTF-8 rune in the input. (so that you can search for phrases containing *any* of the runes in a set of runes, or phrases containing *all of* the runes in a set of runes).

I'll have a think about alternate ways to do the query syntax.

**Comment by sporkmonger** [comment]
Keep in mind that while /8, /16, /24, and /32 are the most common subnet masks, /20, /14, /25, etc are all a thing. I'm not sure that the tokenization you've got there is correct as a result.

I'm pretty sure the go toolchain issue shouldn't a problem in the context of Docker.

**Comment by peterstace** [comment]
Yep, your tokenization algorithm would have to create tokens for all of the subnet masks that you care about (which sounds like it's possible masks).

**Comment by peterstace** [comment]
I came up with the following. It creates a token for each possible mask that the IP address could possibly have.

```
package main

import (
    "net"
)

func Tokens(s string) ([]string, error) {
    _, ipnet, err := net.ParseCIDR(s)
    if err != nil {
        return nil, err 
    }   
    ones, bits := ipnet.Mask.Size()
    var toks []string
    for i := ones; i >= 1; i-- {
        m := net.CIDRMask(i, bits)
        tok := net.IPNet{                                                                                                                                             
            IP:   ipnet.IP.Mask(m),
            Mask: m,
        }   
        toks = append(toks, tok.String())
    }   
    return toks, nil 
}
```

In action:
```
mutation{
  schema{
    ip: string @index(cidr) .
  }
  set{
    _:a <ip> "100.55.22.11/32" .
    _:b <ip> "100.33.81.19/32" .
    _:c <ip> "100.49.21.25/32" .
    _:d <ip> "101.0.0.5/32" .
    _:e <ip> "100.176.2.1/32" .
  }
}
```
```
{
  q(func: customallof(ip, cidr, "100.48.0.0/12")) {
    ip
  }
}
==>
{
  "data": {
    "q": [
      {
        "ip": "100.55.22.11/32"
      },
      {
        "ip": "100.49.21.25/32"
      }
    ]
  }
}
```
```
{
  q(func: customallof(ip, cidr, "100.32.0.0/11")) {
    ip
  }
}
==>
{
  "data": {
    "q": [
      {
        "ip": "100.55.22.11/32"
      },
      {
        "ip": "100.33.81.19/32"
      },
      {
        "ip": "100.49.21.25/32"
      }
    ]
  }
}
```
```
{
  q(func: customallof(ip, cidr, "100.0.0.0/8")) {
    ip
  }
}
==>
{
  "data": {
    "q": [
      {
        "ip": "100.55.22.11/32"
      },
      {
        "ip": "100.33.81.19/32"
      },
      {
        "ip": "100.49.21.25/32"
      },
      {
        "ip": "100.176.2.1/32"
      }
    ]
  }
}
```
```
{
  q(func: customallof(ip, cidr, "101.0.0.0/8")) {
    ip
  }
}
==>
{
  "data": {
    "q": [
      {
        "ip": "101.0.0.5/32"
      }
    ]
  }
}
```

**Comment by peterstace** [comment]
RE go tool chain versions, I don't think docker will automatically save us since a copy of the toolchain/std library will be embedded in both the dgraph binary and in the plugin binary. The docker container doesn't necessarily need to have go installed in it (I'm not sure if it does or not).

**Comment by peterstace** [comment]
Plugins branch merged into master in a574fc6. Note to @sporkmonger , master is undergoing heavy churn right now, so it's probably best to wait for the next official release before giving this a try.
